xayn-ai/src/coi/system.rs

```diff
@@ -382,6 +376,48 @@ impl CoiSystem {
             .collect();
         coi.swap_key_phrases(key_phrases);
     }
+
+    /// Selects the top key phrases from the cois.
+    #[allow(dead_code)]
+    fn select_top_key_phrases<CP: CoiPoint + CoiPointKeyPhrases + CoiPointStats>(
+        &self,
+        cois: &mut [CP],
+        top: usize,
+    ) -> Vec<String> {
+        if top == 0 {
+            return Vec::new();
+        }
+
+        let relevances = self.compute_relevances(cois);
+        let mut relevant_key_phrases = cois
+            .iter_mut()
+            .zip(relevances)
+            .map(|(coi, relevance)| {
+                self.select_key_phrases(coi, &[]);
+                let mut key_phrases = coi
+                    .key_phrases()
+                    .iter()
+                    .map(|key_phrase| (key_phrase.words(), key_phrase.relevance()))
+                    .collect::<Vec<_>>();
+                key_phrases.sort_unstable_by(|(_, this), (_, other)| {
```

Comment:
We can store them already with the order we want in `select_key_phrases`.

Reply:
the new draining algorithm to select the top key phrases is implemented now. two things to note which are different from the original suggestion: instead of
```rust
coiToScore: HashMap<CoiId, Vec<f32>;
phrasesByScore: BTreeMap<f32, Vec<(String, CoiId)>>
```
we have
```rust
coi_to_relevance: HashMap<CoiId, BTreeSet<Relevance>>,
relevance_to_key_phrase: BTreeMap<(Relevance, CoiId), Vec<KeyPhrase>>,
```
where `Relevance` wraps an `f32`. the wrapper is required because the keys of the map must implement `Ord`. the id in the second map is moved from the values to the keys, as this makes iterating & construction & draining more convenient, the desired sorting is still the same.